rasa/core/policies/intent_ted_policy.py

```diff
@@ -0,0 +1,690 @@
+            **model_utilities["meta"],
+        )
+
+
+class IntentTED(TED):
+    def _prepare_label_classification_layers(self, predictor_attribute: Text) -> None:
+        """Prepares layers & loss for the final label prediction step."""
+        self._prepare_embed_layers(predictor_attribute)
+        self._prepare_embed_layers(LABEL)
+
+        self._prepare_dot_product_loss(
+            LABEL, self.config[SCALE_LOSS], loss_layer=layers.MultiLabelDotProductLoss
+        )
+
+    def _prepare_dot_product_loss(
+        self,
+        name: Text,
+        scale_loss: bool,
+        prefix: Text = "loss",
+        loss_layer: tf.keras.layers.Layer = layers.SingleLabelDotProductLoss,
+    ) -> None:
+        self._tf_layers[f"{prefix}.{name}"] = loss_layer(
+            self.config[NUM_NEG],
+            scale_loss,
+            similarity_type=self.config[SIMILARITY_TYPE],
+            model_confidence=self.config[MODEL_CONFIDENCE],
+        )
+
+    @staticmethod
+    def _get_labels_embed(
+        label_ids: tf.Tensor, all_labels_embed: tf.Tensor
+    ) -> tf.Tensor:
+        # instead of processing labels again, gather embeddings from
+        # all_labels_embed using label ids
+
+        indices = tf.cast(label_ids[:, :, 0], tf.int32)
+
+        # Find padding indices. They should have a value -1
+        padding_indices = tf.where(tf.equal(indices, -1))
+
+        # Create a tensor of ones which will serve as updates to original `indices`
+        updates_to_indices = tf.ones((tf.shape(padding_indices)[0]), dtype=tf.int32)
+
+        # Add the tensor of 1s to indices with padding.
+        # So, effectively -1s become 0. This is fine because
+        # we don't change the original label indices but only
+        # make them 'compatible' for the `tf.gather` op below.
+        indices_to_gather = tf.cast(
+            tf.tensor_scatter_nd_add(indices, padding_indices, updates_to_indices),
+            tf.int32,
+        )
+
+        labels_embed = tf.gather(all_labels_embed, indices_to_gather)
+
+        return labels_embed
+
+    def compute_thresholds(
+        self, model_data: RasaModelData, label_ids: np.ndarray
+    ) -> Dict[int, float]:
+        """Compute prediction thresholds for each intent.
+
+        These thresholds are used at inference time to predict
+        whether a query intent is likely or not.
+
+        Args:
+            model_data: Data used during model training.
+            label_ids: Numerical IDs of labels corresponding to data points used during training.
+
+        Returns:
+            Computed thresholds for each intent label present in `label_ids`.
+        """
+        self._training = False
+
+        batch_size = (
+            self.config[BATCH_SIZES]
+            if isinstance(self.config[BATCH_SIZES], int)
+            else self.config[BATCH_SIZES][0]
+        )
+        outputs = self.run_inference(
+            model_data, batch_size=batch_size, output_keys_expected=["similarities"]
```

Comment:
Should this be a constant?

Reply:
It wasn't before in `TEDPolicy` as well, so I've left it for now. Should be done outside of the PR IMO.

Reply:
fair.